core/src/main/java/io/grpc/Attributes.java

```diff
@@ -129,7 +129,18 @@ public boolean equals(Object o) {
       return false;
     }
     Attributes that = (Attributes) o;
-    return Objects.equal(data, that.data);
+    if (data.size() != that.data.size()) {
+      return false;
+    }
+    for (Entry<Key<?>, Object> e : data.entrySet()) {
```

Comment:
This only handles one direction. `that.data` may contain an extra element, but this would incorrectly say they were equal.

Reply:
the sizes are checked above

Reply:
Oh, you're right. kk